xpti/include/xpti/xpti_trace_framework.hpp

```diff
@@ -479,84 +559,930 @@ inline std::string readMetadata(const metadata_t::value_type &MD) {
+  /// trace type is enabled for the stream, the user data and correlation ID are
+  /// communictaed to the subscribers.
+  ///
+  /// @param traceType The trace type to be used for the notification.
+  /// @param userData A pointer to the user data to be included in the
+  /// notification.
+  ///
+  void scopedNotify(uint16_t traceType, const void *userData) {
+    // If tracing is not enabled, don't notify
+    if (!xptiTraceEnabled() || !MData)
       return;
-    switch (m_trace_type) {
-    case signal:
-    case graph_create:
-    case node_create:
-    case edge_create:
-      break;
-    default: {
-      uint16_t close = m_trace_type | 1;
-      xptiNotifySubscribers(m_stream_id, close, m_parent, m_object, m_instance,
-                            m_user_data);
-    } break;
+
+    MTraceType = traceType & 0xfffe;
+    MScopedNotify = true;
+    if (xptiCheckTraceEnabled(MStreamId, traceType) && MTraceType) {
+      MScopedUserData = const_cast<void *>(userData);
+      MScopedCorrelationId = xptiGetUniqueId();
+      // Notify all subscribers with the tracepoint related details
+      xptiNotifySubscribers(MStreamId, traceType, MParentEvent, MTraceEvent,
+                            MScopedCorrelationId, userData);
     }
   }
 
+  /// @brief Adds metadata to the trace event.
+  ///
+  /// This method adds metadata to the trace event if tracing is enabled and the
+  /// trace event is valid. The metadata is added by calling the provided
+  /// callback function with the trace event data.
+  ///
+  /// @param Callback A function that takes a pointer to the trace event data
+  /// and adds the desired metadata.
+  /// @return A reference to this tracepoint_scope_t object.
+  ///
+  tracepoint_scope_t &
+  addMetadata(const std::function<void(xpti::trace_event_data_t *)> &Callback) {
+    if (xptiCheckTraceEnabled(MStreamId, MTraceType) && MTraceEvent) {
+      Callback(MTraceEvent);
+    }
+    return *this;
+  }
+
+  bool isTraceTypeScoped() {
+    if (MTraceType != (uint16_t)xpti::trace_point_type_t::signal &&
+        MTraceType != (uint16_t)xpti::trace_point_type_t::graph_create &&
+        MTraceType != (uint16_t)xpti::trace_point_type_t::node_create &&
+        MTraceType != (uint16_t)xpti::trace_point_type_t::edge_create &&
+        MTraceType != (uint16_t)xpti::trace_point_type_t::queue_create &&
+        MTraceType != (uint16_t)xpti::trace_point_type_t::queue_destroy &&
+        MTraceType != (uint16_t)xpti::trace_point_type_t::diagnostics)
+      return true;
+    else
+      return false;
+  }
+
 private:
-  xpti::trace_event_data_t *m_object, *m_parent;
-  uint8_t m_stream_id;
-  uint16_t m_trace_type;
-  const void *m_user_data;
-  uint64_t m_instance;
+  void selfNotifyEnd() {
+    // selfNotify() is supposed to capture the scope of all the tracepoints we
+    // have enbled; if we need to limit the notifications only to when the trace
+    // scope has valid TLS data, we can add the follwoing IF block if
```

Comment:
I assume you wanted a bit different formatting here

Reply:
Not Doxygen formatting, since we are not describing an API or variable; looks okay to me unless you want to clarify what's different?